cs-vmware/src/main/java/io/cloudslang/content/vmware/services/ClusterComputeResourceService.java

```diff
@@ -0,0 +1,441 @@
+
+        ClusterGroupSpec clusterGroupSpec = new ClusterGroupSpec();
+        clusterGroupSpec.setInfo(clusterHostGroup);
+        clusterGroupSpec.setOperation(ArrayUpdateOperation.REMOVE);
+        clusterGroupSpec.setRemoveKey(vmInputs.getHostGroupName());
+
+        ClusterConfigSpecEx clusterConfigSpecEx = new ClusterConfigSpecEx();
+        clusterConfigSpecEx.getGroupSpec().add(clusterGroupSpec);
+
+        ManagedObjectReference task = connectionResources.getVimPortType().
+                reconfigureComputeResourceTask(clusterMor, clusterConfigSpecEx, true);
+
+        return new ResponseHelper(connectionResources, task).getResultsMap(String.format(SUCCESS_MSG, vmInputs.getClusterName(), task.getValue()), String.format(FAILURE_MSG, vmInputs.getClusterName()));
+    }
+
+    public String listHostGroups(HttpInputs httpInputs, String clusterName, String delimiter) throws Exception {
+        ConnectionResources connectionResources = new ConnectionResources(httpInputs);
+
+        ManagedObjectReference clusterMor = new MorObjectHandler().getSpecificMor(connectionResources, connectionResources.getMorRootFolder(),
+                ClusterParameter.CLUSTER_COMPUTE_RESOURCE.getValue(), clusterName);
+
+        ClusterConfigInfoEx clusterConfigInfoEx = getClusterConfiguration(connectionResources, clusterMor, clusterName);
+
+        List<String> hostGroupNameList = new ArrayList<>();
+        for (ClusterGroupInfo clusterGroupInfo : clusterConfigInfoEx.getGroup()) {
+            if (clusterGroupInfo instanceof ClusterHostGroup) {
+                hostGroupNameList.add(clusterGroupInfo.getName());
+            }
+        }
+        return StringUtilities.join(hostGroupNameList, delimiter);
+    }
+
+    public Map<String, String> createAffinityRule(HttpInputs httpInputs, VmInputs vmInputs,
+                                                  String affineHostGroupName, String antiAffineHostGroupName) throws Exception {
+        ConnectionResources connectionResources = new ConnectionResources(httpInputs);
+
+        ManagedObjectReference clusterMor = new MorObjectHandler().getSpecificMor(connectionResources, connectionResources.getMorRootFolder(),
+                ClusterParameter.CLUSTER_COMPUTE_RESOURCE.getValue(), vmInputs.getClusterName());
+
+        if (ruleExists(getClusterConfiguration(connectionResources, clusterMor, vmInputs.getClusterName()), vmInputs.getRuleName())) {
+            throw new Exception(String.format(RULE_ALREADY_EXISTS, vmInputs.getRuleName()));
+        } else {
+            ClusterVmHostRuleInfo clusterVmHostRuleInfo = getClusterVmHostRuleInfo(getClusterConfiguration(connectionResources, clusterMor, vmInputs.getClusterName()), vmInputs, affineHostGroupName, antiAffineHostGroupName);
+
+            ClusterRuleSpec clusterRuleSpec = new ClusterRuleSpec();
+            clusterRuleSpec.setInfo(clusterVmHostRuleInfo);
+            clusterRuleSpec.setOperation(ArrayUpdateOperation.ADD);
+
```

Comment:
Beggar code. Yes! I did it. :champagne:  We'll talk tomorrow about the duplicates.